rust-escrow/stable-escrow/src/escrow.rs

```diff
@@ -0,0 +1,177 @@
+use near_sdk::{
+    AccountId, Balance, Promise, PromiseOrValue,
+    collections::{LazyOption, UnorderedSet},
+    env, json_types::U128, near_bindgen, serde_json::json,
+};
+
+use near_contract_standards::fungible_token::{
+    FungibleToken,
+    metadata::{FungibleTokenMetadata, FungibleTokenMetadataProvider},
+};
+
+use crate::consts::*;
+use crate::storage::*;
+
+impl Default for Escrow {
+    fn default() -> Self {
+        env::panic_str("Escrow Contract should be initialized before usage")
+    }
+}
+
+#[near_bindgen]
+impl Escrow {
+    #[init]
+    pub fn new(
+        decimals: u8,
+        expires_at: u64,
+        funding_amount_limit: U128,
+        id: String,
+        nep_141_account_id: AccountId,
+        dao_factory_account_id: AccountId,
+        metadata_url: String,
+    ) -> Self {
+        if env::state_exists() {
+            env::panic_str("ERR_ALREADY_INITIALIZED");
+        }
+
+        let id = format!("SA{}", id);
+        let ft_metadata = FungibleTokenMetadata { 
+            spec: "ft-1.0.0".to_string(),
+            name: id.clone(),
+            symbol: id.clone(),
+            icon: None,
+            reference: None,
+            reference_hash: None,
+            decimals: decimals,
+        };
+
+        let mut token = FungibleToken::new(b"t".to_vec());
+        token.total_supply = funding_amount_limit.0;
+
+        Self {
+            deposits: UnorderedSet::new(b"d".to_vec()),
+            ft: token,
+            ft_metadata: LazyOption::new(b"m".to_vec(), Some(&ft_metadata)),
+            metadata: Metadata {
+                expires_at,
+                funding_amount_limit: funding_amount_limit.0,
+                unpaid_amount: funding_amount_limit.0,
+                nep_141_account_id,
+                dao_factory_account_id,
+                metadata_url
+            }
+        }
+    }
+
+    /**
+     * Called on ft_transfer_callback only
+     * Contract NEP141 balance is kept in the NEP141 contract
+     * Sender balances are kept in the balances map
+     * Total sender balances must match the contract NEP141 balance, minus fees
+     *
+     * Transfer self NEP141 in exchange of the stable NEP141 amount as a receipt
+     */
+    #[private]
+    pub fn deposit(&mut self, sender_id: AccountId, amount: Balance) {
+        if !self.is_deposit_allowed() {
+            env::panic_str("ERR_DEPOSIT_NOT_ALLOWED");
+        }
+
+        if amount > self.get_unpaid_amount() {
+            env::panic_str("ERR_DEPOSIT_NOT_ALLOWED");
+        }
+
+        self.ft.internal_deposit(&sender_id, amount);
+        self.deposits.insert(&sender_id);
```

Comment:
@susguzman should we keep track of this now that we have `self.ft.internal_deposit(&sender_id, amount);`?

Reply:
Yes, because we the structure that the NEP141 uses is non-iterable, but if I found a better way to track it I will remove it.